Adds an optional consumer that's called whenever a value is set.

lewisd32 opened the pull request:
This allows for a usage pattern like:
```Java
ValueHolder usernameAlias = new ValueHolder();
ValueHolder passwordAlias = new ValueHolder();

new DslParams(args,
    new RequiredParam("username").setConsumer(usernameAlias),
    new RequiredParam("password").setConsumer(passwordAlias));

String username = testContext.usernames.resolve(usernameAlias.get());
String password = testContext.passwords.resolve(passwordAlias.get());

driver().login(username, password);
```

Or even something like this (which I'd like to implement in [unacceptable](https://github.com/unacceptable/unacceptable)):
```Java
ValueHolder username = new ValueHolder(testContext.usernames);
ValueHolder password = new ValueHolder(testContext.password);

new DslParams(args,
    new RequiredParam("username").setConsumer(username),
    new RequiredParam("password").setConsumer(password));

driver().login(username.get(), password.get());
```

Compared to the current way:
```Java
DslParams params = new DslParams(args,
    new RequiredParam("username")
    new RequiredParam("password"));

String username = testContext.usernames.resolve(params.value("username"));
String password = testContext.passwords.resolve(params.value("password");

driver().login(username, password);
```

I think the new way is more consise, and elliminates duplication of parameter names, particularly in the case of an `OptionalParam` without a default, where one generally ends up doing:
```Java
if (params.hasValue("name"))
{
    String name = params.value("name");
    // do something with name
}
```

mikeb01 commented:
Looks good to me.  I'd like @ajsutton to have a quick look first before merging.

mikeb01 commented:
I'm wondering if it should be BiConsumer<String, String> so that the name of the parameter can be passed as well?  There may be cases where you want the same consumer to take multiple parameters?

lewisd32 commented:
That sounds like a good idea to me.  I'll make that change.

lewisd32 commented:
I kept the existing `setConsumer(Consumer<String> consumer)` method, because it's useful for passing in method handles (eg, `list::add`), but I've changed the `consumer` internally to be a `BiConsumer` and added `setConsumer(BiConsumer<String, String> consumer)`.

ajsutton commented:
I'm not sure I see the value in having a separate ValueHolder object rather than just retrieving the value from DslParams.  For example:

```java
    DslParams params = new DslParams(args, 
        new RequiredParam("username"),
        new RequiredParam("password"));

    driver().login(params.value("username"), params.value("password"));
```

is a lot more succinct than:

```java
    ValueHolder username = new ValueHolder();
    ValueHolder password = new ValueHolder();
    
    new DslParams(args,
        new RequiredParam("username").setConsumer(username),
        new RequiredParam("password").setConsumer(password));
    
    driver().login(username.get(), password.get());
```

But I definitely see the value in being able to apply transformations. So I wonder if it would be better to just add a setTransformer(BiFunction<String, String> transform) method:

```java
    DslParams params = new DslParams(args, 
        new RequiredParam("username")
            .setTransformer((name, value) -> testContext.usernames.resolve(value)),
        new RequiredParam("password"));

    driver().login(params.value("username"), params.value("password"));
```

Probably with a Function<String, String> override so we can just say setTransformer(testContext.usernames::resolve).

Apart from being a bit more succinct, it preserves access to the range of valueAs methods. It doesn't avoid repeating the parameter name, but I think if you're buying into using DslParams parameter names are intentionally just strings and the param can always be retrieved once and stored in a variable.   For optional params the valueAsOptional accessor tends to be very effective:

```java
    DslParams params = new DslParams(args, new OptionalParam("value"));
    params.valueAsOptional("value").ifPresent(value -> doSomething(value));
```

What do you think?

ajsutton commented:
Actually, one other thing worth mentioning is that you can currently do:

```java
    SimpleDslParam username = new RequiredParam("username");
    SimpleDslParam password = new SimpleDslParam("password");
    new DslParams(args, username, password);

    driver.login(username.getValue(), password.getValue());
```

It's not the way I'd write it, but it lets you avoid repeating the param name as a string if you particularly want that.  Probably most useful when combined with the setTransformer suggested above.

lewisd32 commented:
It hadn't occurred to me to pull the params out into fields.  I think that might address some of my concerns, though it does make it a bit harder for someone reading the method (someone who's writing tests) to reason about the order of parameters, etc.
`valueAsOptional` (which I didn't know about) would also make some code a bit clearer by reducing the number of times the string parameter name needs to be used.

I kind of like the `BiTransformer` idea, but when used with `.value()` it would only return `String`.  Resolving names from aliases are only one of the places the consumer/transformer pattern would be useful. There are a number of cases where I'd like to transform the string into an object of some sort. The "parameter combining" example on the wiki is a good example of where that might be handy.
```Java
fixAPI.placeMassOrder("FTSE100", "bid: 10@38", "bid: 11@39", "ask: 13@44", "ask: 14@45");
```
Within the method, one could have something like:
```Java
SimpleDslParam bidParam = new OptionalParam("bid").setAllowMultipleValues().setTransformer(BidTransformer::transform);
DslParams params = new DslParams(args, bidParam, ...);
PriceAndQuantity[] bids = (PriceAndQuantity[]) bidParam.transformedValues();
```
But it would require a cast based on whatever type the transformer returns, which is a bit messy. I'm not sure the generics to fix that could be added without either breaking backwards compatibility, or introducing the consumer.

The consumer way:
```Java
ValueHolder bids = ValueHolder.multipleValues(BidTransformer::transform);

DslParams params = new DslParams(args,
    new OptionalParam("bid").setAllowMultipleValues().setConsumer(bids),
    ...);

// bids.get() returns List<PriceAndQuantity>
```
(I haven't tried compiling the code I've typed here, so my apologies if I missed something.)
It's not perfect, since it does duplicate stating that the param allows multiple values. However, it is typesafe.

Some other examples of conversions to non-string types I've been using:
* Parsing durations (eg. "11 seconds") for asserting that cookies on an API are set with an appropriate expiry.
* Parsing a date range (eg. `"createdDate: between testStart and afterUserCreated"`)
* Converting strings like "OK" and "NOT_FOUND" to 200 and 400 for asserting expected status codes.

ajsutton commented:
That's a reasonable argument and I don't think supporting that style of usage detracts from directly using params.value methods.

It also occurs to me that this could be really useful with a builder:

```java
    OrderBuilder builder = new OrderBuilder();
    DslParams params = new DslParams(args,
        new RequiredParam("instrument").setConsumer(builder::instrument),
        new OptionalParam("side").setConsumer(builder::side),
        new OptionalParam("expectedStatus").setDefault("ACCEPTED"));

    driver.placeOrder(builder.newInstance());
    driver.waitForStatus(params.value("expectedStatus"));
```

I'll probably still look at implementing transformers when I get a chance - they'd always return String since the dsl is deliberately very stringy and then the various params.valueAs methods can be used to retrieve values as long, int etc.

Thanks for the patch and for taking the time to discuss the changes.
